File: apps/api/app/models/runtime.py
"""SQLAlchemy ORM 模型 - Skill、MCP、Memory、Knowledge 等运行时资源。"""

from datetime import datetime
from sqlalchemy import String, DateTime, ForeignKey, Text, Integer, Boolean, Float
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base


class SkillModel(Base):
    """Skill 表 - 对应 SKILL.md 文件。"""
    __tablename__ = "skills"

    skill_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), ForeignKey("organizations.org_id"), nullable=False, index=True)
    team_id: Mapped[str | None] = mapped_column(String(64), ForeignKey("teams.team_id"), nullable=True)
    agent_id: Mapped[str | None] = mapped_column(String(64), ForeignKey("agents.agent_id"), nullable=True)
    scope: Mapped[str] = mapped_column(String(32), nullable=False)  # bundled/organization/team/agent
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    content: Mapped[str] = mapped_column(Text, default="")
    file_path: Mapped[str | None] = mapped_column(String(512), nullable=True)  # SKILL.md 文件路径
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class AgentSkillPolicyModel(Base):
    """Agent Skill 授权策略表。"""
    __tablename__ = "agent_skill_policies"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    agent_id: Mapped[str] = mapped_column(String(64), ForeignKey("agents.agent_id"), nullable=False, index=True)
    skill_id: Mapped[str] = mapped_column(String(64), ForeignKey("skills.skill_id"), nullable=False, index=True)
    allowed: Mapped[bool] = mapped_column(Boolean, default=True)


class MCPServerModel(Base):
    """MCP Server 表。"""
    __tablename__ = "mcp_servers"

    server_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), ForeignKey("organizations.org_id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    transport: Mapped[str] = mapped_column(String(16), default="http")
    url: Mapped[str] = mapped_column(String(512), nullable=False)
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class MCPToolModel(Base):
    """MCP Tool 表。"""
    __tablename__ = "mcp_tools"

    tool_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    server_id: Mapped[str] = mapped_column(String(64), ForeignKey("mcp_servers.server_id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    input_schema: Mapped[str] = mapped_column(Text, default="{}")
    risk_level: Mapped[str] = mapped_column(String(16), default="low")
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class AgentMCPPolicyModel(Base):
    """Agent MCP 授权策略表。"""
    __tablename__ = "agent_mcp_policies"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    agent_id: Mapped[str] = mapped_column(String(64), ForeignKey("agents.agent_id"), nullable=False, index=True)
    server_id: Mapped[str] = mapped_column(String(64), ForeignKey("mcp_servers.server_id"), nullable=False, index=True)
    allowed: Mapped[bool] = mapped_column(Boolean, default=True)


class MemoryModel(Base):
    """Memory 记忆表。"""
    __tablename__ = "memories"

    memory_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    agent_id: Mapped[str] = mapped_column(String(64), ForeignKey("agents.agent_id"), nullable=False, index=True)
    memory_type: Mapped[str] = mapped_column(String(32), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    summary: Mapped[str] = mapped_column(Text, default="")
    confidence: Mapped[float] = mapped_column(Float, default=0.5)
    source: Mapped[str] = mapped_column(String(64), default="user")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class ModelProviderModel(Base):
    """模型供应商配置表。"""
    __tablename__ = "model_providers"

    provider_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), ForeignKey("organizations.org_id"), nullable=False, index=True)
    provider_key: Mapped[str] = mapped_column(String(64), nullable=False)
    display_name: Mapped[str] = mapped_column(String(128), nullable=False)
    base_url: Mapped[str] = mapped_column(String(512), nullable=False)
    api_key_encrypted: Mapped[str] = mapped_column(Text, default="")
    api_key_masked: Mapped[str] = mapped_column(String(64), default="")
    models_json: Mapped[str] = mapped_column(Text, default="[]")
    default_model: Mapped[str] = mapped_column(String(128), default="")
    is_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class BackgroundAgentModel(Base):
    """后台 Agent 配置表。"""
    __tablename__ = "background_agents"

    config_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), ForeignKey("organizations.org_id"), nullable=False, index=True)
    agent_type: Mapped[str] = mapped_column(String(32), nullable=False)  # memory/mcp_health/workflow_monitor/queue_governor
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    interval_seconds: Mapped[int] = mapped_column(Integer, default=300)
    status: Mapped[str] = mapped_column(String(16), default="idle")
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)


class SkillEvaluationModel(Base):
    """Skill 使用评价表。"""
    __tablename__ = "skill_evaluations"

    evaluation_id: Mapped[str] = mapped_column(String(64), primary_key=True)
    org_id: Mapped[str] = mapped_column(String(64), ForeignKey("organizations.org_id"), nullable=False, index=True)
    agent_id: Mapped[str] = mapped_column(String(64), ForeignKey("agents.agent_id"), nullable=False, index=True)
    skill_id: Mapped[str] = mapped_column(String(64), ForeignKey("skills.skill_id"), nullable=False, index=True)
    session_id: Mapped[str | None] = mapped_column(String(64), ForeignKey("sessions.session_id"), nullable=True)
    user_input: Mapped[str] = mapped_column(Text, default="")
    assistant_output: Mapped[str] = mapped_column(Text, default="")
    status: Mapped[str] = mapped_column(String(32), default="pending")  # pending/evaluated/applied/rejected
    score: Mapped[float | None] = mapped_column(Float, nullable=True)
    failure_reason: Mapped[str] = mapped_column(Text, default="")
    improvement_suggestion: Mapped[str] = mapped_column(Text, default="")
    proposed_skill_patch: Mapped[str] = mapped_column(Text, default="")
    applied: Mapped[bool] = mapped_column(Boolean, default=False)
    created_by: Mapped[str] = mapped_column(String(64), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

File: apps/web/types/runtime.ts
/** Runtime 相关类型定义。 */

/** Skill 实体 */
export interface Skill {
  skill_id: string;
  name: string;
  description: string;
  scope: string;
}

/** Skill 使用评价 */
export interface SkillEvaluation {
  evaluation_id: string;
  org_id: string;
  agent_id: string;
  skill_id: string;
  session_id: string | null;
  user_input: string;
  assistant_output: string;
  status: string;
  score: number | null;
  failure_reason: string;
  improvement_suggestion: string;
  proposed_skill_patch: string;
  applied: boolean;
  created_by: string;
  created_at: string;
}

/** MCP Server */
export interface MCPServer {
  server_id: string;
  name: string;
  transport: string;
  url: string;
}

/** MCP Tool */
export interface MCPTool {
  tool_id: string;
  name: string;
  description: string;
  risk_level: string;
}

/** Memory 条目 */
export interface MemoryItem {
  memory_id: string;
  memory_type: string;
  summary: string;
  confidence: number;
}

/** Session 条目 */
export interface SessionItem {
  session_id: string;
  status: string;
  compact_summary: string;
}

/** Context Bundle */
export interface ContextBundle {
  total_estimated_tokens: number;
  need_compaction: boolean;
  sections: Array<{ name: string; content: string; estimated_tokens: number }>;
}

/** 模型供应商 */
export interface ModelProvider {
  provider_id: string;
  provider_key: string;
  display_name: string;
  base_url: string;
  api_key_masked: string;
  models: string[];
  default_model: string;
  is_enabled: boolean;
}

/** LLM 调用日志 */
export interface LLMCallLog {
  call_id: string;
  provider: string;
  model: string;
  status: string;
  prefix_hash: string;
}

/** 后台 Agent */
export interface BackgroundAgentItem {
  config_id: string;
  org_id: string;
  agent_type: string;
  enabled: boolean;
  interval_seconds: number;
  status: string;
}

/** Skill 创建请求 */
export interface CreateSkillRequest {
  actor_user_id: string;
  org_id: string;
  scope: string;
  team_id?: string;
  agent_id?: string;
  content: string;
}

/** MCP 创建请求 */
export interface CreateMCPServerRequest {
  actor_user_id: string;
  org_id: string;
  name: string;
  transport: string;
  url: string;
}

/** Memory 创建请求 */
export interface CreateMemoryRequest {
  actor_user_id: string;
  agent_id: string;
  memory_type: string;
  content: string;
  summary: string;
  confidence: number;
  source: string;
}

/** Model Provider 创建请求 */
export interface CreateModelProviderRequest {
  actor_user_id: string;
  org_id: string;
  provider_key: string;
  display_name: string;
  base_url: string;
  api_key: string;
  models: string[];
  default_model: string;
}
